ingestion/src/kldata.rs

```diff
@@ -0,0 +1,279 @@
+}
+
+fn parse_obs(csv_body: Lines<'_>, columns: &[ObsinnId]) -> Result<Vec<ObsinnObs>, Error> {
+    let mut obs = Vec::new();
+
+    for row in csv_body {
+        let (timestamp, vals) = {
+            let mut vals = row.split(',');
+
+            let raw_timestamp = vals
+                .next()
+                .ok_or_else(|| Error::Parse("empty row in kldata csv".to_string()))?;
+
+            // TODO: timestamp parsing needs to handle milliseconds and truncated timestamps?
+            let timestamp = DateTime::parse_from_str(raw_timestamp, "%Y%m%d%H%M%S")
+                .map_err(|e| Error::Parse(e.to_string()))?
+                .with_timezone(&Utc);
+
+            (timestamp, vals)
+        };
+
+        for (i, val) in vals.enumerate() {
+            let col = columns[i].clone(); // Should we do some smart bounds-checking??
+
+            // TODO: parse differently based on param_code?
+
+            obs.push(ObsinnObs {
+                timestamp,
+                id: col,
+                value: val.parse().map_err(|_| {
+                    Error::Parse(format!("value {} could not be parsed as float", val))
+                })?,
+            })
+        }
+    }
+
+    Ok(obs)
+}
+
+pub fn parse_kldata(msg: &str) -> Result<(usize, ObsinnChunk), Error> {
+    // parse the first two lines of the message as meta, and csv column names,
+    // leave the rest as an iter over the lines of csv body
+    let (station_id, type_id, message_id, columns, csv_body) = {
+        let mut lines = msg.lines();
+
+        let lines_err = || Error::Parse("kldata message contained too few lines".to_string());
+
+        let (station_id, type_id, message_id) = parse_meta(lines.next().ok_or_else(lines_err)?)?;
+        let columns = parse_columns(lines.next().ok_or_else(lines_err)?)?;
+
+        (station_id, type_id, message_id, columns, lines)
+    };
+
+    Ok((
+        message_id,
+        ObsinnChunk {
+            observations: parse_obs(csv_body, &columns)?,
+            station_id,
+            type_id,
+        },
+    ))
+}
+
+// TODO: rewrite such that queries can be pipelined?
+// not pipelining here hurts latency, but shouldn't matter for throughput
+pub async fn filter_and_label_kldata(
+    chunk: ObsinnChunk,
+    conn: &mut PooledPgConn<'_>,
+    param_conversions: Arc<HashMap<String, (String, i32)>>,
+    permit_table: Arc<RwLock<PermitTable>>,
+) -> Result<Vec<Datum>, Error> {
+    let query_get_obsinn = conn
+        .prepare(
+            "SELECT timeseries \
+                FROM labels.obsinn \
+                WHERE nationalnummer = $1 \
+                    AND type_id = $2 \
+                    AND param_code = $3 \
+                    AND lvl = $4 \
+                    AND sensor = $5",
+        )
+        .await?;
+
+    let mut data = Vec::with_capacity(chunk.observations.len());
+
+    for in_datum in chunk.observations {
+        // get the conversion first, so we avoid wasting a tsid if it doesn't exist
```

Comment:
Agree, I think this is a mistake on my part. Though I find myself leaning toward putting param_id in the filter label instead of element_id. What is the justification for using element_id?

Reply:
Could we just make the ingestor label with the obsinn label for now, and we deal with the element_id later? Its great that it can convert to element_id as this may be something we need for confident and also for when we do manage to label with a filter label (I can try to talk to Ketil a bit about this today).

Reply:
Yes, I agreed with that. I asked a different question.

Reply:
My suggestion is to not label with filter at all for now. The idea with filter is to make it more "user friendly" as it is what most people would actually want from say Frost. But collapsing typeid (aka partly different time resolutions) has some caveats we need to deal with I think.

Reply:
The convention set by stinfosys seems to be to use integer ids in the main tables and names only in conversion tables, so we ought to have a good reason to diverge from that.

Reasons I see for param_id:

1. Prevents corruption/inconsistency from typos and name changes
2. Stores and indexes more efficiently (not that big a deal since this table is not that big)
3. Keeps consistency with what we're doing elsewhere

Reasons I see for element_id:

1. Saves Frost having to do a conversion

Based on this, paramid makes more sense to me, but there may be history I'm unaware of

Reply:
Ok, so is the question actually if we want paramid in the obsinn label? That I am ok with, that is what happens when its sent through kvalobs now. It probably makes sense to get away from code, and my understanding from Vegar is we may even be able to ask Søren to send it with paramid instead of paramcode.

Reply:
Not the obsinn label, the filter label, obsinn should have param code, because that's what it comes in with

Reply:
> My suggestion is to not label with filter at all for now

I don't see any point kicking the can down the road on that. The database is not useful without the filter label, and having to run a script to make labels later is more work than just getting this right now.

Reply:
I've replaced filter label with a met label that uses param_id and type_id, as per our discussion